src/topics/TopicEditModal.js

```diff
@@ -0,0 +1,129 @@
+// @flow strict-local
+import React, { useState, useContext, useEffect } from 'react';
+import { Modal, View } from 'react-native';
+import type { Node } from 'react';
+import styles, { ThemeContext, BRAND_COLOR, createStyleSheet } from '../styles';
+import { updateMessage } from '../api';
+import type { Auth, GetText, Stream } from '../types';
+import { fetchSomeMessageIdForConversation } from '../message/fetchActions';
+import ZulipTextIntl from '../common/ZulipTextIntl';
+import ZulipTextButton from '../common/ZulipTextButton';
+import Input from '../common/Input';
+
+type Props = $ReadOnly<{|
+  topicModalProviderState: {
+    visible: boolean,
+    topic: string,
+    streamId: number,
+  },
+  auth: Auth,
+  zulipFeatureLevel: number,
+  streamsById: Map<number, Stream>,
+  _: GetText,
+  closeEditTopicModal: () => void,
+|}>;
+
+export default function TopicEditModal(props: Props): Node {
+  const { topicModalProviderState, closeEditTopicModal, auth, zulipFeatureLevel, streamsById, _ } =
+    props;
+
+  const { visible, topic, streamId } = topicModalProviderState;
+
+  const [topicName, onChangeTopicName] = useState();
+
+  useEffect(() => {
+    onChangeTopicName(topic);
+  }, [topic]);
+
+  const { backgroundColor } = useContext(ThemeContext);
+
+  const modalStyles = createStyleSheet({
+    wrapper: {
+      flex: 1,
+      justifyContent: 'center',
+      alignItems: 'center',
+    },
+    modal: {
+      justifyContent: 'flex-start',
+      backgroundColor,
+      padding: 15,
+      shadowOpacity: 0.5,
+      shadowColor: 'gray',
+      shadowOffset: {
+        height: 5,
+        width: 5,
+      },
+      shadowRadius: 5,
+      borderRadius: 5,
+      width: '90%',
+    },
+    buttonContainer: {
+      flexDirection: 'row',
+      justifyContent: 'flex-end',
+    },
+    titleText: {
+      fontSize: 18,
+      lineHeight: 21,
+      color: BRAND_COLOR,
+      marginBottom: 10,
+      fontWeight: 'bold',
+    },
+  });
+
+  const handleSubmit = async () => {
+    if (topicName === '') {
+      return;
+    }
+    const messageId = await fetchSomeMessageIdForConversation(
+      auth,
+      streamId,
+      topic,
+      streamsById,
+      zulipFeatureLevel,
+    );
+    if (messageId == null) {
+      throw new Error(
+        _('No messages in topic: {streamAndTopic}', {
+          streamAndTopic: `#${streamsById.get(streamId)?.name ?? 'unknown'} > ${topic}`,
+        }),
+      );
+    }
+    await updateMessage(auth, messageId, {
+      propagate_mode: 'change_all',
+      subject: topicName,
+      ...(zulipFeatureLevel >= 9 && {
+        send_notification_to_old_thread: true,
+        send_notification_to_new_thread: true,
+      }),
+    });
+    closeEditTopicModal();
```

Comment:
Let's add some `try`/`catch`/`finally` logic here:

- I think everything except the `closeEditTopicModal()` should go in the `try`.
- The `closeEditTopicModal()` should go in the `finally`.
- Then a `showToast` or `showErrorAlert` can go in the `catch`, so the user knows the operation has failed.